### ISSUE TITLE
TileSet::duplicate(false) breaks original TileSet resource

### ISSUE THREAD

**Issue by FluffyJay1** [godotengine/godot#80076]
### Godot version

4.2.dev2

### System information

Godot v4.2.dev2 - Windows 10.0.19044 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1080 (NVIDIA; 31.0.15.3179) - AMD Ryzen 5 3600 6-Core Processor (12 Threads)

### Issue description

The mere invocation of `TileSet::duplicate(false)` breaks the original TileSet resource. In particular it looks like the TileSetSources of the original TileSet are getting removed, but I can only speculate.

### Example A

In this example both TileMaps are using the same TileSet resource, but one of them has the script attached; invoking `TileSet::duplicate(false)` causes both of them to lose their tiles.

https://github.com/godotengine/godot/assets/5732925/0b362a46-e88d-43ce-a028-58daceca267e

### Example B

This example is the same as above, except the first tilemap gets the newly duplicated TileSet, and the second one keeps the original TileSet:

https://github.com/godotengine/godot/assets/5732925/3cc57bfb-b692-4685-b786-6a0ad1619e9d

### Steps to reproduce

1. Create a TileSet resource and attach it to a TileMap
2. At runtime call `duplicate(false)` on the TileSet resource
3. Observe that the TileMap is now blank

### Minimal reproduction project

[TileSetDuplicate.zip](https://github.com/godotengine/godot/files/12210276/TileSetDuplicate.zip)

**Comment by KoBeWi** [comment]
AFAIK tile sources are tied to a TileSet. Doing `duplicate(false)` duplicates the TileSet and assigns the old sources to the new TileSet, so they are transferred. Not sure if this is fixable.

**Comment by groud** [comment]
Yes, as Kobewi's says, TileSet have internal resources that needs to be duplicated with the TileSet if you want it to keep working. I guess we could mention that somewhere in the the docs.

**Comment by AThousandShips** [comment]
This is documented partially [here](https://docs.godotengine.org/en/latest/classes/class_tilesetsource.html#class-tilesetsource) and [here](https://docs.godotengine.org/en/latest/classes/class_tileset.html#class-tileset-method-add-source) but should be added more clearly about duplication too

**Comment by FluffyJay1** [comment]
My intuition tells me shallow duplicating a TileSet (or any resource for that matter) shouldn't adversely modify the original resource merely because of an implementation detail. Shouldn't the TileSetSources be duplicated as well, no matter what?

**Comment by AThousandShips** [comment]
It would be pretty involved and complicated to force partially deep duplication on `TileSet`, and I don't think it's a good idea

We *could* make it error if you try shallow duplication as well as documentation, it should be as simple as:
```cpp
	virtual Ref<Resource> duplicate(bool p_subresources = false) const override {
		ERR_FAIL_COND_V_MSG(!p_subresources, Ref<Resource>(), "TileSet cannot be duplicated without duplicating subresources");
		return Resource::duplicate(true);
	}
```
